lib/elixir/lib/kernel.ex

```diff
@@ -3069,6 +3081,38 @@ defmodule Kernel do
     end
   end
 
+  # Called as ensure_evaled(list, {0, []}). Note acc is reversed.
+  def ensure_evaled(list, acc) do
+    :lists.mapfoldl(fn
+      {:|, meta, [head, tail]}, acc ->
+        {head, acc} = ensure_evaled_element(head, acc)
+        {tail, acc} = ensure_evaled_tail(tail, acc)
+      {{:|, meta, [head, tail]}, acc}
+      elem, acc ->
+        ensure_evaled_element(elem, acc)
+    end, acc, list)
+  end
+
+  defp ensure_evaled_element(elem, acc)
+       when is_number(elem) or is_atom(elem) or is_binary(elem) do
+    {elem, acc}
+  end
+  defp ensure_evaled_element(elem, acc) do
+    ensure_evaled_var(elem, acc)
+  end
+
+  defp ensure_evaled_tail(elem, acc) when is_list(elem) do
+    ensure_evaled(elem, acc)
+  end
+  defp ensure_evaled_tail(elem, acc) do
+    ensure_evaled_var(elem, acc)
+  end
+
+  defp ensure_evaled_var(elem, {i, ast}) do
```

Comment:
I guess this `i` here is index, not integer, I think it'd be better to spell fully.